src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+            }
+
+            var lessThanToken = EatToken(SyntaxKind.LessThanToken);
+            var saveTerm = _termState;
+            _termState |= TerminatorState.IsEndOfFunctionPointerParameterList;
+            var types = _pool.AllocateSeparated<FunctionPointerParameterOrReturnTypeSyntax>();
+
+            try
+            {
+                while (true)
+                {
+                    var modifiers = _pool.Allocate<SyntaxToken>();
+                    try
+                    {
+                        // Modifiers will be checked for validity in later steps
+                        while (IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword)
+                        {
+                            modifiers.Add(EatToken());
+                        }
+
+                        var parameterType = ParseTypeOrVoid();
+                        types.Add(SyntaxFactory.FunctionPointerParameterOrReturnType(modifiers, parameterType));
+
+                        if (skipBadFunctionPointerParameterListTokens() == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+
+                        Debug.Assert(CurrentToken.Kind == SyntaxKind.CommaToken);
+                        types.AddSeparator(EatToken(SyntaxKind.CommaToken));
+                    }
+                    finally
+                    {
+                        _pool.Free(modifiers);
+                    }
+                }
+
+                var greaterThanToken = EatToken(SyntaxKind.GreaterThanToken);
+                var funcPointer = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanToken, types, greaterThanToken);
+                funcPointer = CheckFeatureAvailability(funcPointer, MessageID.IDS_FeatureFunctionPointers);
+                return funcPointer;
+            }
+            finally
+            {
+                _termState = saveTerm;
+                _pool.Free(types);
+            }
+
+            PostSkipAction skipBadFunctionPointerParameterListTokens()
+            {
+                CSharpSyntaxNode? tmp = null;
+                Debug.Assert(types.Count > 0);
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
+                    types,
+                    isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                    abortFunction: p => p.CurrentToken.Kind == SyntaxKind.CommaToken || p.IsTerminator(),
+                    expected: SyntaxKind.CommaToken);
+            }
+        }
+
+        private bool IsFunctionPointerStart()
+            => CurrentToken.Kind == SyntaxKind.DelegateKeyword && PeekToken(1).Kind == SyntaxKind.AsteriskToken;
+
+        private bool IsPossibleCallingConvention()
+            => !IsPossibleFunctionPointerParameterListStart(CurrentToken) &&
+               IsPossibleFunctionPointerParameterListStart(PeekToken(1)) &&
```

Comment:
i don't get the need for these two checks.  why is `CurrentToken.Kind == SyntaxKind.IdentifierToken || IsPredefinedType(CurrentToken.Kind)` not a sufficient check?

Reply:
To try and have a little nicer parse trees. My guess here is that, if there's no list start in sight, we'll have better error recovery if we don't blindly parse this given token as a calling convention. For example, the user could be in the process of adding a cast, something like `(delegate*MyFunctionCall()`, where the end goal is `(delegate*<void>)MyFunctionCall()`.

---
In reply to: [348850685](https://github.com/dotnet/roslyn/pull/39879#discussion_r348850685) [](ancestors = 348850685)

Reply:
i see.  
1. that sounds reasonable.
2. imo we should doc this.  code in the parser that doesn't do the simplest/minimal sort of check possible tends to be difficult to understand later on.  I looked at this and said "i cannot reconcile what this doing against the grammar".  My reading was correct, but your justification was also correct.  The difference between the two is where comments are useful.  Thanks! :)